File: PriceFeed/src/priceFeed.py
#!/bin/python3

import datetime
import eth_abi
import json
import urllib.request
import sha3
import sys

root                = ''
inFolder            = '{}iexec_in/'.format(root)
outFolder           = '{}iexec_out/'.format(root)
determinismFilePath = '{}determinism.iexec'.format(outFolder)
callbackFilePath    = '{}callback.iexec'.format(outFolder)
errorFilePath       = '{}error.iexec'.format(outFolder)

datasetEnvVar       = 'IEXEC_DATASET_FILENAME'

DEFAULT_APIKEY = '65FFBB07-6119-4A1E-80DC-F46B4FF9EE2F';
# DEFAULT_APIKEY = '69CC0AA9-1E4D-4E41-806F-8C3642729B88';
# DEFAULT_APIKEY = 'D2C881D6-0BBF-4EFE-A572-AE6DB379D43E';
# DEFAULT_APIKEY = 'FB7B2516-70A1-42D8-8702-292F29F19768';

if __name__ == '__main__':

	# GET API KEY
	try:
		with open('{}{}'.format(inFolder, os.environ[datasetEnvVar]), 'r') as keyFile:
			apikey = keyFile.read()
	except:
		apikey = DEFAULT_APIKEY

	# GET PARAMS
	asset_id_base  = sys.argv[1] if len(sys.argv) > 1 else None
	asset_id_quote = sys.argv[2] if len(sys.argv) > 2 else None
	power          = sys.argv[3] if len(sys.argv) > 3 else None
	time           = sys.argv[4] if len(sys.argv) > 4 else None

	# FORMAT TIME
	try:
		time = datetime.datetime.utcfromtimestamp(int(time))
	except:
		try:
			time = datetime.datetime.fromisoformat(time)
		except:
			time = datetime.datetime.utcnow()

	try:
		# REQUEST
		host     = 'rest.coinapi.io'
		fragment = '&'.join('{}={}'.format(k,v) for k,v in {
			'time': time.isoformat()[0:23]+"Z"
		}.items())
		url      = 'https://{}/v1/exchangerate/{}/{}?{}'.format(host, asset_id_base, asset_id_quote, fragment)
		headers  = { 'X-CoinAPI-Key': apikey }
		data     = urllib.request.urlopen(urllib.request.Request(url, headers=headers)).read()
		results  = json.loads(data)

		# if 'error' in results:
		# 	raise urllib.request.HTTPError(url, 0, "invalid results", headers, None)

		timestamp = int(datetime.datetime.fromisoformat(results['time'][0:23]).timestamp()*1000)
		details   = '-'.join([ results['asset_id_base'], results['asset_id_quote'], power ]);
		value     = round(float(results['rate']) * 10 ** int(power))

		# ENCODE CALLBACK
		callback = eth_abi.encode_abi(['uint256', 'string', 'uint256'], [timestamp, details, value]).hex()
		with open(callbackFilePath, 'w') as callbackFile:
			callbackFile.write('0x{}'.format(callback))

		# GENERATE DETERMINISM
		determinism = sha3.keccak_256()
		determinism.update(bytes.fromhex(callback))
		with open(determinismFilePath, 'w') as determinismFile:
			determinismFile.write('0x{}'.format(determinism.hexdigest()))

		print('- Success: {} {} {}'.format(timestamp, details, value));

	except urllib.request.HTTPError as httperror:
		# STORE ERROR
		with open(errorFilePath, 'w') as errorFile:
			errorFile.write(str(httperror))

		# GENERATE DETERMINISM
		determinism = sha3.keccak_256()
		determinism.update(str(httperror).encode())
		with open(determinismFilePath, 'w') as determinismFile:
			determinismFile.write('0x{}'.format(determinism.hexdigest()))

File: PriceFeed/src/priceFeed.js
const https   = require('https');
const ethers  = require('ethers');
const fs      = require('fs');

const root                = 'iexec_out';
const determinismFilePath = `${root}/determinism.iexec`;
const callbackFilePath    = `${root}/callback.iexec`;
const errorFilePath       = `${root}/error.iexec`;


/*****************************************************************************
 *                                   TOOLS                                   *
 *****************************************************************************/
const sleep = (ms) => {
	return new Promise(resolve => setTimeout(resolve, ms));
}

const cat = (path) => {
	try { return fs.readFileSync(path).toString(); } catch (e) { return null; }
}

/*****************************************************************************
 *                                  CONFIG                                   *
 *****************************************************************************/

// coin api key
const APIKEY = '65FFBB07-6119-4A1E-80DC-F46B4FF9EE2F';
// const APIKEY = '69CC0AA9-1E4D-4E41-806F-8C3642729B88';
// const APIKEY = 'D2C881D6-0BBF-4EFE-A572-AE6DB379D43E';
// const APIKEY = 'FB7B2516-70A1-42D8-8702-292F29F19768';

// --- version 0: using an aggregate ---
const VERSION = 0;

// --- version 1: using trade history ---
// const VERSION = 1;
// const EXCHANGE = 'BITSTAMP';
// const EXCHANGE = 'BINANCE';

// random delay
const WAIT_MIN = parseInt(process.env.WAIT_MIN) || 0; // in ms
const WAIT_MAX = parseInt(process.env.WAIT_MAX) || 0; // in ms

/*****************************************************************************
 *                                 ARGUMENTS                                 *
 *****************************************************************************/

var [ asset_id_base, asset_id_quote, power, time ] = process.argv.slice(2).map(s => s.toUpperCase());

if (/^\d*$/.test(time)) { time = new Date(parseInt(time)*1000); } // evm timestamp
else if (time)          { time = new Date(time);                } // any other format
else                    { time = new Date();                    } // not value → now

/*****************************************************************************
 *                                HTTP QUERY                                 *
 *****************************************************************************/
let path = undefined;

switch (VERSION)
{
	case 0:
	{
		const fragment = Object.entries({
			'time': time.toISOString(),
		}).filter(([k,v]) => v).map(([k,v]) => `${k}=${v}`).join('&');

		path = `/v1/exchangerate/${asset_id_base}/${asset_id_quote}?${fragment}`;
		break;
	}
	case 1:
	{
		const time_start = new Date(time.getTime() - 60*1000); // T - 60s
		const time_end   = new Date(time.getTime() +  0*1000); // T +  0s

		const fragment = Object.entries({
			'time_start': time_start.toISOString(),
			'time_end':   time_end.toISOString(),
		}).filter(([k,v]) => v).map(([k,v]) => `${k}=${v}`).join('&');

		path = `/v1/trades/${EXCHANGE}_SPOT_${asset_id_base}_${asset_id_quote}/history?${fragment}`;
		break;
	}
}

const query = {
	method: 'GET',
	port:   443,
	host:   'rest.coinapi.io',
	path:   path,
	headers: { 'X-CoinAPI-Key': cat(`/iexec_in/${process.env.IEXEC_DATASET_FILENAME}`) || APIKEY },
};

/*****************************************************************************
 *                                  EXECUTE                                  *
 *****************************************************************************/
new Promise(async (resolve, reject) => {

	const delay = (WAIT_MAX-WAIT_MIN) * Math.random() + WAIT_MIN;
	console.log(`- Waiting for ${delay} ms.`);
	await sleep(delay);

	console.log(`- Calling API ${query.host}${query.path}`);
	let chunks = [];
	let request = https.request(query, res => {
		res.on('data', (chunk) => {
			chunks.push(chunk);
		});
		res.on('end', () => {
			if (chunks.length)
			{
				resolve(chunks.join(''));
			}
			else
			{
				reject(`[HTTP ERROR]\nstatusCode: ${res.statusCode}`);
			}
		});
	});
	request.on('error', reject);
	request.end();
})
.then(data => {
	let results = JSON.parse(data.toString());

	if (results.error !== undefined)
	{
		throw new Error(results.error);
	}

	let timestamp = undefined;
	let details   = undefined;
	let value     = undefined;

	switch (VERSION)
	{
		case 0:
		{
			timestamp = new Date(results.time).getTime();
			details   = [ results.asset_id_base, results.asset_id_quote, power].join('-');
			value     = Math.round(results.rate * 10**power);
			break;
		}
		case 1:
		{
			if (results.length == 0)
			{
				throw new Error('missing data');
			}
			const average = ((w,v) => v/w)(...results.reduce((acc, trade) => [acc[0] + trade.size, acc[1] + trade.price * trade.size], [0,0]));
			timestamp = time.getTime();
			details   = [ asset_id_base, asset_id_quote, power].join('-');
			value     = Math.round(average * 10**power);
			break;
		}
	}

	if (isNaN(timestamp) || isNaN(value))
	{
		throw new Error("invalid results");
	}

	var iexeccallback = ethers.utils.defaultAbiCoder.encode(['uint256', 'string', 'uint256'], [timestamp, details, value]);
	var iexecconsensus = ethers.utils.keccak256(iexeccallback);
	fs.writeFile(callbackFilePath,    iexeccallback , (err) => {});
	fs.writeFile(determinismFilePath, iexecconsensus, (err) => {});

	console.log(`- Success: ${timestamp} ${details} ${value}`);
})
.catch(error => {
	fs.writeFile(
		errorFilePath,
		error.toString(),
		(err) => {}
	);
	fs.writeFile(
		determinismFilePath,
		ethers.utils.solidityKeccak256(['string'],[error.toString()]),
		(err) => {}
	);
	console.log(error.toString());
});
